[클린코드 리액트 3기 정수현] 계산기 구현

sasha1107 opened the pull request:
## 구현 의도

- 뷰에 표시되는 값을 하나의 state(`screenValue`)로 관리하고 `parseExpression` 함수로 `screenValue`로부터 연산식을 해석(연산자와 피연산자를 분리)하도록 했습니다. 
- 숫자의 입력가능 자릿수는 확장성(변동 가능성)을 위해 상수로 관리하였습니다.
- 최대한 간단하게 구현하여 가독성을 중시하였습니다.
 
## 질문

- 전체적인 로직을 한 곳에서 파악할 수 있도록 하기 위함과 props를 내리는 depth가 모두 1 depth로 props drilling이 심하지 않다고 판단되어 모든 이벤트 핸들러 함수를 컴포넌트의 최상단에서 선언하고 props로 내려주는 식으로 구현했는데 `App.tsx`의 코드가 다소 길어진 감이 있는 것 같기도 해서 이에 대한 의견과 개선이 필요하다면 어떤 식으로 개선하면 좋을지 의견을 듣고 싶습니다.

- 연산식을 해석하는 `parseExpression` 함수에서 [String.prototype.match()](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/String/match) 메서드의 반환값의 타입이 RegExpMatchArray | null 타입으로 추론되어 타입을 리터럴 타입의 유니온으로 더 좁혀주기 위해 타입 단언을 사용하였습니다. 
https://github.com/next-step/react-calculator/blob/3544aa19815b45248a9634e30e97e69fec99360b/src/utils/parseExpression.ts#L13
    - 이 부분에서 타입 단언 사용을 지양하고 타입스크립트에 의해 해당 타입으로 추론하게 할 수 있는 방법이 있을까요?

sasha1107 committed:
Created project with vite

sasha1107 committed:
feat: 계산기 구현

zereight commented on index.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<html lang="ko">
```
어?! 이런 꼼꼼함 챙기기 쉽지 않은데!!

zereight commented on src/utils/index.ts:
```diff
@@ -0,0 +1,4 @@
+import calculate from './calculate';
+import parseExpression from './parseExpression';
+
+export { calculate, parseExpression };
```
오 배럴파일도 만드셨군요!

수현님의 어떤의도(?)에서 배럴파일을 만들게되셨는지 궁금합니다! (진짜 단순궁금한거임)

sasha1107 replied:
개인적으로 import 문을 그룹화하여 짧고 가독성 좋게 만들기 위해 사용하는 것을 선호하는 편입니다. 😀

zereight commented on src/utils/parseExpression.ts:
```diff
@@ -0,0 +1,18 @@
+import { OPERATORS } from '@/constants';
+import type { OperatorType } from '@/types';
+
+const parseExpression = (expression: string) => {
```
utils인데 계산기 도메인인 operator에 종속적이라는 생각이 드네용!

util이려면 도메인과 좀 더 떨어진 연관성을 가지면 좋겠다는 생각이 들어요!

zereight commented on src/utils/parseExpression.ts:
```diff
@@ -0,0 +1,18 @@
+  );
+  const numbers = expression
+    .split(operatorRegex)
+    .map((num) => parseInt(num, 10))
```
parseInt와 Number의 차이는 무엇일까요?!

sasha1107 replied:
정확한 차이를 알지 못한 채로 사용하고 있었네요😅

|`parseInt`|`Number`|
|:-:|:-:|
|문자열 타입의 매개변수를 정수로 리턴해주는 함수|숫자 객체를 생성하는 생성자 함수|
|숫자와 문자 중 숫자만 인식하여 숫자로 리턴, 숫자가 없을 경우 `NaN`|숫자로 이루어진 것만 숫자로 리턴,숫자가 아닌게 섞여있을 경우 `NaN`|
|소수점 X|소수점도 인식|

일반 문자(숫자가 아닌 것)가 섞이지 않아 기능 상으로는 동일하나, 소수점을 인식할 필요는 없기 때문에 `parseInt`로 통일하는 방향으로 수정하였습니다!

zereight commented on src/utils/parseExpression.ts:
```diff
@@ -0,0 +1,18 @@
+import type { OperatorType } from '@/types';
+
+const parseExpression = (expression: string) => {
+  const operatorRegex = new RegExp(
+    `[${OPERATORS.map((operator) => `\\${operator.label}`).join('')}]`,
+    'g'
+  );
```
보통 중첩된 template literals는 살짝 가독성이 떨어지는 편인거같긴해요! ㅎㅎ (취향차이 일수도)

zereight commented on src/utils/parseExpression.ts:
```diff
@@ -0,0 +1,18 @@
+    .split(operatorRegex)
+    .map((num) => parseInt(num, 10))
+    .filter((num) => !Number.isNaN(num));
+  const operator = expression.match(operatorRegex) as OperatorType[] | null;
```
String.match와 Regexp.exec의 차이도 알아보시면 살짝 재밌을수도?!

sasha1107 replied:
||String.match|Regexp.exec|
|:-:|:-:|:-:|
|반환값|배열|배열|
|형식|문자열.match(정규식)|정규식.exec(문자열)|
|정규식에 `g` 플래그를 붙여서 함수를 실행|문자열에서 해당하는 부분을 전부 배열에 넣어서 return|처음 일치하는 부분 하나만 배열에 넣어서 return한다. 단. 정규식에 캡처 그룹이 포함되어 있을 경우 해당하는 값을 같이 반환|

```js
const peach = '복숭아빙수 복숭아김치 복숭아스무디';
const exp2 = /복숭아(..)/g;

console.log(peach.match(exp2));
//["복숭아빙수", "복숭아김치", "복숭아스무"] 출력

console.log(exp2.exec(peach));
//["복숭아빙수", "빙수"] 출력
```
> **캡쳐 그룹**: 은 괄호로 둘러싼 영역

- - -

새롭게 알게된 내용이네요! 좋은 정보 감사합니다.😄😄

zereight replied:
저도 직접 측정해보지는 않았지만, 미세하게 성능차이도 있다고 합니당 ㅎㅎ 참고하세요!

zereight commented on src/utils/parseExpression.ts:
```diff
@@ -0,0 +1,18 @@
+    .split(operatorRegex)
+    .map((num) => parseInt(num, 10))
+    .filter((num) => !Number.isNaN(num));
+  const operator = expression.match(operatorRegex) as OperatorType[] | null;
```
> 이 부분에서 타입 단언 사용을 지양하고 타입스크립트에 의해 해당 타입으로 추론하게 할 수 있는 방법이 있을까요?

`OPERATORS`를 `as const`로 잘 선언해주셔서,
조건문 같은걸로, 타입을 좁혀줄 수 있을거같아요!

"타입 좁히기"는 typescript한테 그렇게 밖에 추론하지 못하는 상황을 강제해주면 됩니당 ㅎ_ㅎ

```ts
  const arr = OPERATORS.map(({ label }) => label);

  if (!operator) {
    throw new Error("Invalid expression");
  }

  if (arr.includes(operator[0])) {
    return { numbers, operator: operator[0] };
  }
```

zereight commented on vite.config.ts:
```diff
@@ -0,0 +1,13 @@
+  plugins: [react()],
+  resolve: {
+    alias: {
+      '@': path.resolve(__dirname, './src'),
```
캬

zereight commented on src/utils/calculate.ts:
```diff
@@ -0,0 +1,25 @@
+  divide,
+};
+
+const calculate = (numbers: number[], operator: OperatorType) => {
```
numbers가 정말 2개의 number만 받는거라면
`number[]`보다는 `[number, number]`로 타입을 주는게더, 의도가 명확하고 안전한 타입정의라고 할 수 있겠네요!!

zereight commented on src/utils/calculate.ts:
```diff
@@ -0,0 +1,25 @@
+    console.error('Invalid operator');
+    return 0;
+  }
+  return OPERATOR_ACTIONS[operateAction](num1, num2);
```
넘 좋네용!

zereight commented on src/constants/index.ts:
```diff
@@ -0,0 +1,37 @@
+    label: '+',
+    action: 'sum',
+  },
+] as const;
+
+const ERROR_MESSAGE = {
+  MAX_NUMBER_LENGTH: `숫자는 최대 ${MAX_NUMBER_LENGTH}자리까지 입력 가능합니다.`,
+  MAX_NUMBER_COUNT: `숫자는 최대 ${MAX_NUMBER_COUNT}개까지 입력 가능합니다.`,
+  NO_NUMBER: '숫자를 먼저 입력해주세요.',
+};
+
```
`as const`를 사용하고 하지않고의 특별한 기준을 가지시고 계신게있는지 궁금하네용!!

sasha1107 replied:
리터럴 타입으로 추론 범위를 좁혀서 선언한 변수로부터 파생한 타입을 얻어야 할 때만 사용하는 편입니다.

예시)
```ts
const OPERATORS = [
  {
    label: '/',
    action: 'divide',
  },
  {
    label: 'X',
    action: 'multiply',
  },
  {
    label: '-',
    action: 'subtract',
  },
  {
    label: '+',
    action: 'sum',
  },
] as const;
type OperatorType = (typeof OPERATORS)[number]['label']; // '/' | 'X' | '-' | '+'
```
상수라면 `as const`를 사용하는 것이 권장되는 패턴일지 궁금합니다.

zereight replied:
`as const` 를 사용하면, 타입이 명확해지고 내부 속성들에 readonly 가 부여되어서 객체의 오염을 막을 수 있다는 장점이 있습니다 ㅎㅎ

예를들어서 javascript로만 개발을 하시면서, 객체를 상수화할때
`Object.freeze` 또는 `seal` 을 사용해서 객체 속성의 혹시나 싶은 재할당으로 인한 오염을 방지하기도 하는데요. (불변성)

`as const`를 사용하면 readonly가 부여되어서, 객체를 수정하지 못하니까 좀 더 엄격하게 관리한다는 점에서 상수면 보통 `as const`를 써주는 것 같습니다 ㅎㅎ

그리구 정말로 관련된 목적으로만 기술을 사용하시려는것은 너무 좋은 개발자의 모습인거같습니다!

zereight commented on src/components/Screen.tsx:
```diff
@@ -0,0 +1,9 @@
+  return <h1 id='total'>{value}</h1>;
+};
+
+export default Screen;
```
간단한 컴포넌트라도, 의미있게 컴포넌트를 분리한거 저는 찬성입니다!

하지만 `Screen` 이름보다 더 좋은게 있지 않을까라는 생각이 드네용!

`Screen`은 뭔가 하나의 페이지를 담당하는 컴포넌트 느낌이 든달까!!

zereight commented on src/components/Operations.tsx:
```diff
@@ -0,0 +1,38 @@
+
+const Operations = ({
+  handleClickOperator,
+  handleClickCalculate,
```
`<button onClick` 처럼, react에서는 `onEvent` 형식으로 이벤트 콜백인자를 내려주고 있어요.

그래서 보통 prop으로 이벤트콜백을 내려줄때에는 prop명을 `on~~`형식을 사용하는게 관행인거같습니다 ㅎㅎ.

prop으로 내리지 않는다면 `handle~`을 붙여도 되고, 안붙여도 되고요 ㅎㅎ.

제가 담당하는 프로덕트도

```tsx

const handleClick = () => {}

return <button onClick={handleClick}>aa</button>

```
요런식으로 컨벤션이 정해진게 있답니다 ㅎㅎ

zereight commented on src/components/AC.tsx:
```diff
@@ -0,0 +1,14 @@
+type ACProps = {
+  handleClickAllClear: () => void;
+};
+const AC = ({ handleClickAllClear }: ACProps) => {
```
컴포넌트명이 모두 대문자라서 살짝 어색한 느낌이있네요! (줄여서 그런가)

보통 컴포넌트이름은 `PascalCase`로 쓰는게 관행입니다!

요거는 호불호 갈릴 수 있으니, 추후에 다른 리뷰어분의 의견도 들어보는게 좋을거같습니다. ㅎㅎ

sasha1107 replied:
lint warning이 발생하고 있어 수정하는게 좋을 것 같아 수정하여 반영하였습니다 :)

> Fast refresh only works when a file only exports components. Use a new file to share constants or functions between components.

zereight commented on src/App.tsx:
```diff
@@ -0,0 +1,88 @@
+  };
+  const handleClickCalculate = () => {
+    inputNum.current = '';
+    if (numbers.length < 2 || !operator) {
```
매직넘버 상수화 요것도 해주시면 깔끔할거같습니다!!

zereight commented on src/App.tsx:
```diff
@@ -0,0 +1,88 @@
+  };
+  const handleClickAllClear = () => {
+    inputNum.current = '';
+    setScreenValue('0');
```
ScreenValue 초기값이 `'0'` 이라서, 여기도 `'0'`군요!

같은 의미라면, 이것도 상수화 해놓는게 추후 유지보수에 좋을거같습니다!

초기값을 변경해야하는 상황이 온다면, 수정을 2번해야하니까요!
그리고 왜 `'0'` 으로 초기화하는지에 대한, 의도도 명확히 나타낼 수 있고요 ㅎㅎ

zereight commented on src/App.tsx:
```diff
@@ -0,0 +1,88 @@
+      setScreenValue(`${digit}`);
+      return;
+    }
+    const newNum = Number(`${inputNum.current}${digit}`).toString();
```
어떤건 template literal로, 어떤건 toString()으로 문자열화하고 계시네요!

별다른 이유가 없다면, 통일하는게 좋을거같습니다.

물론 그렇게 되면, 중첩된 template literal이 되는데요.

```ts
const newNum = Number(`${inputNum.current}${digit}`);
const newNumAsStr = `${newNum}`;
```
이렇게 자료형을 쉽게 알 수 있게, 구별해주어도 상관없을것 같습니다. 

(요 피드백은 취향을 좀 탈것 같으니, 요런 피드백도 있구나~하고 넘기셔도 될거같아요)

zereight commented on src/App.tsx:
```diff
@@ -0,0 +1,88 @@
+  const [screenValue, setScreenValue] = useState('0');
+  const inputNum = useRef('');
+  const { numbers, operator } = parseExpression(screenValue);
+  const operatorLabels = OPERATORS.map(({ label }) => label as string);
```
UI에 표시되는 값을 `screenValue` 하나의 상태로 관리하려다보니, 

타입을 다시 string으로 바꾸는 부분도 생긴거같구, `screenValue`를 위해서, 문자열 분리하여 연산하고 하는 과정이 쵸큼 복잡해보이긴하는데요!

연산에 필요한 상태는 직관적인 계산기 로직처럼, firstNum, operator, secondNum으로 나누고
`screenValue`는 UI에 렌더링되는 값이니까, 그때그때 파생된 상태(Derived State)로 관리하는게 더 직관적이고 테스트를 작성하기 쉽지 않을까? 생각이 들긴했습니다!

하지만 요건, 계산기 같이 통상적으로 알고있는 도메인(첫번쨰숫자+연산자+두번째숫자)이라 
더 그렇게 느껴지는것 같아요!
하나의 역할을 하는 상태로 줄이고, 코드를 단순화하시려는 노력 매우 인상깊었습니다!

zereight commented on src/App.tsx:
```diff
@@ -0,0 +1,88 @@
+  const inputNum = useRef('');
+  const { numbers, operator } = parseExpression(screenValue);
+  const operatorLabels = OPERATORS.map(({ label }) => label as string);
+  const isError = screenValue === '오류';
```
업데이트가 잦은, UI 텍스트를 기준으로 로직이 짜여지는건 좋지않은 방향같아요!

물론, 이 미션이 간단하니까 편의상 이렇게 구현하신거 알고있지만!

다음 미션때에는 한번더 고려해보아요!

zereight commented on src/constants/index.ts:
```diff
@@ -0,0 +1,37 @@
+const MAX_NUMBER_LENGTH = 3;
+const MAX_NUMBER_COUNT = 2;
+
+const DIGITS = Array.from({ length: 10 }, (_, i) => 9 - i);
```
이것도 뭔가 type으로 잘 정의되었으면 좋지 않았을까 싶은데요!

길이가 길지 않으니, 
`const DIGITS = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] as const;` 으로 해도 좋지 않았을까 생각합니다!

zereight commented on src/components/Operations.tsx:
```diff
@@ -0,0 +1,38 @@
+      <button
+        className='operation'
+        onClick={handleClickCalculate}
+        role='button'
```
요거 `role='button'`이 원래 있었나용?

tag가 button인데, role까지 button으로 주어서 불필요한 속성선언인거 같아요!

불필요하게 중복된 선언은 오히려 가독성을 해칠 수 있을거같아요!

sasha1107 replied:
테스트 코드를 작성 해볼까하고 `getByRole`을 사용하고자 추가해놓았는데 삭제하는 것을 깜빡했네요!!
현재 코드에서는 불필요한 선언이므로 삭제하였습니다.

zereight reviewed (changes requested):
안녕하세요. 수현님!! 반갑습니당.

수현님이 작성해주신 코드를 보면서, 많은 생각을 하셨구나가 느껴졌습니당 ㅎㅎ.

react, ts에 어느정도 익숙해지신분 같아서, 조금 더 자세하게 피드백드렸어요!

물론, 계산기 미션이 프로세스 맛보기용이니까 
모든 피드백을 반영해주실 필요는 없으시고.

매직넘버 상수화? 정도만 반영해주셔도 approve 드리도록 하겠습니다!

다른 피드백들은 다음 미션에 알잘딱하여 녹이셔도 될거같아요! (질문주셔도 좋구용!)

질문답변

> 전체적인 로직을 한 곳에서 파악할 수 있도록 하기 위함과 props를 내리는 depth가 모두 1 depth로 props drilling이 심하지 않다고 판단되어 모든 이벤트 핸들러 함수를 컴포넌트의 최상단에서 선언하고 props로 내려주는 식으로 구현했는데 App.tsx의 코드가 다소 길어진 감이 있는 것 같기도 해서 이에 대한 의견과 개선이 필요하다면 어떤 식으로 개선하면 좋을지 의견을 듣고 싶습니다.

아뇨! 저는 props drilling에서 전혀 악취가 나지 않습니다! 1depth 라서요. 

> 연산식을 해석하는 parseExpression 함수에서 [String.prototype.match()](https://developer.mozilla.org/ko/docs/Web/JavaScript/Reference/Global_Objects/String/match) 메서드의 반환값의 타입이 RegExpMatchArray | null 타입으로 추론되어 타입을 리터럴 타입의 유니온으로 더 좁혀주기 위해 타입 단언을 사용하였습니다.

ㅎㅎ 피드백 남겨놨습니다!

sasha1107 committed:
refactor: 매직넘버 상수로 분리

sasha1107 committed:
refactor: 형변환 방식 통일

sasha1107 committed:
refactor: Digit Type 지정

sasha1107 committed:
rename: 컴포넌트명 변경

- AC -> AllClear
  컴포넌트명을 uppercase로 작성하여 발생한 lint warning 수정
  warning: Fast refresh only works when a file only exports components. Use a new file to share constants or functions between components.
- Screen -> ScreenPanel
  Screen이 페이지를 담당하는 느낌을 줄 수 있다는 리뷰 반영

sasha1107 committed:
refactor: calculate 파라미터 타입 튜플로 수정

sasha1107 committed:
remove: 불필요한 속성 선언 삭제

sasha1107 committed:
refactor: props명 수정

sasha1107 committed:
refactor: String.match -> Regexp.exec로 수정

zereight reviewed (approved):
오!! 수현님 수고하셨습니다!

다음미션때 수현님의 코드가 더 기대되네요!!!
